### PR TITLE
move shape to Node base

### PR DESCRIPTION
**Pull request by wconstab**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: torch/csrc/lazy/ts_backend/ts_node.cpp [hunk 1]
```diff
@@ -27,11 +27,7 @@ const Shape& GetShapeFromTsValue(const Value& value) {
 
 void TsNodeSetShapeDeferred(
     NodePtr node, const std::function<Shape()>& shape_fn) {
-  if (auto tsnode = std::dynamic_pointer_cast<TsNode>(node)) {
-    tsnode->SetShapeDeferred(shape_fn);
-    return;
-  }
-  throw std::runtime_error("Expected TsNode but could not dynamic cast");
+  throw std::runtime_error("SetShapeDeferred is not supported. IRs calling this aren't used anyway");
```
Comment (Krovatkin) [inline torch/csrc/lazy/ts_backend/ts_node.cpp:30]: why can't we remove it then?

**Review by Krovatkin — APPROVED** [review]
Looks reasonable. 
I was wondering if we could keep `sizes()` and size()` virtual still, so we can override them for `torch::lazy::DimensionNode`s to throw an error?

File: torch/csrc/lazy/ts_backend/ts_node.cpp (hunk 1 above)
Comment (wconstab) [reply torch/csrc/lazy/ts_backend/ts_node.cpp:30]: yea, it should be removed. I had a final commit that removed it and also broke something else, so i just dropped that one and moved to land the working part first

File: torch/csrc/lazy/core/ir.h [hunk 2]
```diff
@@ -158,6 +99,15 @@ class TORCH_API Node {
...
   }
 
+  // Shapes are computed by meta kernels and represent the static shape for the current graph,
+  // or in the case of dynamic ops, the 'upper bound' shape based on inferring the maximal output
+  // shape given the input shapes
+  // Retrieves a vector of shapes each representing the shape of one output tensor for this op
+  c10::ArrayRef<Shape> shapes() const { return shapes_; }
+
+  // Retrieves the shape of the output at a given index.
+  const Shape& shape(size_t output_index = 0) const;
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/csrc/lazy/core/ir.h:109]: yea, i can leave it as virtual

[the remark this answers is not in the document]
Comment (wconstab) [reply torch/csrc/lazy/core/ir.h]: oops. So I had you review 6 commits, but I intended to drop the last commit.  What I did now is just dropped the last commit and left the original first 5 commits.  The last commit that I dropped did a few things
 - made shape non-virtual
 - got rid of SetShapeDeferred
 - moved owning of Shape obj to Node base, and updated all the constructors
 
 This commit broke something and I never had time to figure it out, so I'm landing just the part that works first to unblock others

**Pull request merged**